internal/usbgadget/config.go

```diff
@@ -177,36 +185,57 @@ func (u *UsbGadget) Init() error {
 
 	u.udc = udcs[0]
 
-	err := u.configureUsbGadget(false)
+	err := u.configureUsbGadget(UsbResetAlways)
 	if err != nil {
 		return u.logError("unable to initialize USB stack", err)
 	}
 
 	return nil
 }
 
-func (u *UsbGadget) UpdateGadgetConfig() error {
+func (u *UsbGadget) UpdateGadgetConfig(resetUsbMode UsbResetMode) error {
 	u.configLock.Lock()
 	defer u.configLock.Unlock()
 
 	u.loadGadgetConfig()
 
-	err := u.configureUsbGadget(true)
+	err := u.configureUsbGadget(resetUsbMode)
 	if err != nil {
 		return u.logError("unable to update gadget config", err)
 	}
 
 	return nil
 }
 
-func (u *UsbGadget) configureUsbGadget(resetUsb bool) error {
-	return u.WithTransaction(func() error {
-		u.tx.MountConfigFS()
-		u.tx.CreateConfigPath()
-		u.tx.WriteGadgetConfig()
-		if resetUsb {
-			u.tx.RebindUsb(true)
+func (u *UsbGadget) configureUsbGadget(resetUsbMode UsbResetMode) error {
+	f := func(resetUsbAfter bool) func() error {
+		return func() error {
+			u.tx.MountConfigFS()
+			u.tx.CreateConfigPath()
+			u.tx.WriteGadgetConfig()
+			if resetUsbAfter {
+				u.tx.RebindUsb(true)
+			}
+			return nil
 		}
+	}
+
+	// initial attempt to configure the gadget
+	err := u.WithTransaction(f(resetUsbMode == UsbResetAlways))
+	if err == nil {
 		return nil
-	})
+	}
+
+	// if the initial attempt failed, try to configure the gadget again with the resetUsb flag
+	if resetUsbMode == UsbResetOnDemand {
+		u.log.Warn().Err(err).Msg("initial attempt to configure the gadget failed, resetting USB gadget then trying again")
+		// NOTES: do not use the RebindUsb method here because it will block the transaction
```

Comment:
...do not use the _public_ RebindUsb method... 

I completely misread this the first time ;)